Improve EPICSMotors and EPICSActuator classes

RafaelLyra8 opened the pull request:
- Create a simplified EPICSMotor class with the standart EPICSMotors fields already configured (This simplifies the YAML configuration file of every motor, reducing from 9 parameters to be be configured to just 1)
- Fix _wait_actuator function that wasn't working on both EPICSMotors and EPICSActuator

marcus-oscarsson commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,90 @@
+    def init(self):
+        """Initialization method"""
+        super(EPICSActuator, self).init()
+        gevent.spawn(self._watch)
```
Do you really need to perform this polling of the value like this cant it be done via the EPICS command/channels ?

RafaelLyra8 replied:
I think using the EPICS command would really be better, I'll study how to implement it.

marcus-oscarsson commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,90 @@
+    def _watch(self):
+        """Watch Actuator current value and update it on the UI."""
+        while True:
+            gevent.sleep(0.25)
```
You are "observing" `self._nominal_value` in the "background" but at the same time your the one setting `self._nominal_value`. I'm perhaps missing something but I don't think this would be needed ?

RafaelLyra8 replied:
It's really not necessary, I saw that the update_value function in the AbstractActuator already checks if the PV value has changed or not. I'll remove it.

beteva replied:
Not only unnecessary, but may induce errors, as update_value will not see the change of the value and not emit the signal.

marcus-oscarsson commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,90 @@
+        """Override AbstractActuator method."""
+        self.set_channel_value(self.ACTUATOR_VAL, value)
+
+    def set_value(self, value, timeout=0):
```
Are you sure that you need to overload `set_value`, its perhaps sufficient to overload, `wait_ready`.
If you look at:

   https://github.com/mxcube/mxcubecore/blob/bc5381a8ce804193c5cc925eee6b9cc51d284474/mxcubecore/HardwareObjects/abstract/AbstractActuator.py#L179C17-L179C41

You can see that it does basically the same as your `set_value`.

You could add a `self.update_state(self.STATES.BUSY)` in your `_set_value` and implement the `__wait_actuator_task` in `wait_ready`, or similair.

I would use a event mechanism instead of the spawn (if its available)

beteva replied:
Agree with @marcus-oscarsson.

rhfogh replied:
set_value was written to be generic, calling _set_value for the actual setting. If you need something additional done I would even say to do the additional stuff and then call the AbstractActuator function. Better not duplicate that if not absolutely necessary.

marcus-oscarsson commented on mxcubecore/HardwareObjects/LNLS/LNLSDiffractometer.py:
```diff
@@ -0,0 +1,345 @@
+        # Armazena clique do usuário
+        self.last_centred_position[0] = x
+        self.last_centred_position[1] = y
+        print(f"Última posição clicada pelo usuário: x={x}, y={y}")
```
Be careful with using prints, these are not logged/saved anywhere. They are also in Spanish. I don't see it as an issue as its in your site specific code, but you may want to change it to English.

RafaelLyra8 replied:
This will be removed in a future version, I only focused on the EpicsMotor and EpicsActuator classes and left the old code for the LNLSDiffractometer that is being used. I'll update it in another PR.

marcus-oscarsson commented on mxcubecore/HardwareObjects/LNLS/LNLSDiffractometer.py:
```diff
@@ -0,0 +1,345 @@
+        self.sampy = None
+
+    def init(self):
+        """
```
I think you can remove these old documentation placeholders :)

marcus-oscarsson commented:
Very nice, I'm a bit puzzled by how the the `EPICSActuator` handles updates, maybe there is a good reason for it that I don't see.

beteva commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,90 @@
+
+    def init(self):
+        """Initialization method"""
+        super(EPICSActuator, self).init()
```
As we dropped the python2 support, you may want to use python3 only syntax

beteva commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,90 @@
+            gevent.sleep(0.15)
+            cur = time.time()
+            if (cur - start) > timeout:
+                raise TimeoutError
```
May I suggest again the use of pyton3 syntax:
with gevent.Timeout( timeout):
            while hasnt_arrived(setpoint):
                time.sleep(0.15)

RafaelLyra8 commented:
The update are currently being handled with the watcher method (self._watch), but as you mentioned before, this would be better if it was handled by the Command.

beteva commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSMotor.py:
```diff
@@ -0,0 +1,79 @@
+    def _wait_actuator(self, value):
+        """Override EPICSActuator method."""
+        while not self.done_movement() or self.hasnt_arrived(value):
+            time.sleep(0.25)
```
You've removed the timeout mechanism from the wait command, Does that mean that you always want to wait forever?

RafaelLyra8 replied:
No, I will implement this. But, as of right now, the timeout isn't being set as I didn't really understand how can I configure the devices timeout, is it in the YAMLs?

beteva replied:
There is no rule of how to do it. In general we assume that timeout of None is waiting forever, 0 means return immediately without waiting the execution to finish and a float value is in s to wait a reasonable amount of time. In general timeout is a parameter of the method. If you want, you can add default value in the configuration file of each device (the motors in this case), What I do in general is to set default value to None in the call of the method and when calling the method, set it to some value if needed. You can have a look in the ExporterMotor.py to see how the timeout is implemented for this kind of motors.

beteva reviewed (commented):
Nice job.
It is a bit unfortunate that you implement your diffractometer now, with the GenericDiffractometer. If you can wait for another few weeks, you may want to base it on the new AbstractDiffarctometer abstract class.

RafaelLyra8 commented:
> Nice job. It is a bit unfortunate that you implement your diffractometer now, with the GenericDiffractometer. If you can wait for another few weeks, you may want to base it on the new AbstractDiffarctometer abstract class.

This is only an initial version, based on the current class that is currently being used, I will refactor its code in the next few weeks, and it shouldn't be a problem to wait for the new Abstract Diffractometer class as the new version of mxcube will take some months before going for production.

marcus-oscarsson commented:
Great, nicely done. Could you please rebase and then you can merge ?

**edit**: For some reason the changes you made, also made it look like most comments where addressed but there still seems to be quite a few remaining right ?, Did you plan to look a those as well ?

rhfogh commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,70 @@
+            return False
+        return not np.isclose(readback, setpoint, rtol=self.unit, atol=self.unit)
+
+    def wait_ready(self, timeout):
```
The superclass function has the timeout parameter optional. You cannot make it mandatory in a subclass

beteva replied:
The code did not change, as it has been requested by @rhfogh.

rhfogh commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSMotor.py:
```diff
@@ -0,0 +1,97 @@
+        # Timeout tolerance
+        timeout += 10
+        try:
+            with gevent.Timeout(timeout, exception=TimeoutError):
```
What would be the reason for not using the existing machinery, with ready_event?

pedrobenetton replied:
I tried to implement ready_event with our `self.done_movement()` and `self.hasnt_arrived()` functions in both EPICSActuator and EPICSMotor. I think it is being properly used now.

rhfogh replied:
I do not really understand. The existing wait_ready function has only two lines:
```
        with Timeout(timeout, RuntimeError("Timeout waiting for status ready")):
            self._ready_event.wait(timeout=timeout)
```
The only thing you need to  do is to ensure that self._ready_event is set and cleared appropriately by the code elsewhere, you do not need wait_ready to do anything. If you have chosen to bypass that and re-implement the event waiting functionality afresh, what is the reason for that? What do you need that cannot be done in the existing framework, or why is it not possible for the rest of the code to set and clear that event appropriately?

beteva replied:
Hi @rhfogh, I do not know the requirements of the epics motors, but it might be that you should alway give some (timeout=0 might not be allowed). @RafaelLyra8, could you, please, confirm this.

RafaelLyra8 replied:
I decided to set the timeout automatically in order to simplify the configuration file and because the EPICS Motor Record, which is the standard EPICS abstraction that I want this class to encompass, contains the setpoint, readback and velocity parameters with which I can discover the motion duration and I added some tolerance time of motion.
I also reimplemented the wait_ready class because the EPICS Motor Record provide the information if the motion has finished, that's why I implemented the done_movement method, and in order to encompass cases that this logic might be faulty I added another validation which verify if the readback is the same as the setpoint, considering some tolerance.
We could leave the class using the standard implementation of timeout but because of the motor record standards, I wanted to create this class in order to simplify the configuration. This simplification made motor XML that used to have 15 lines without considering timeouts and limits to YAMLs with 7 with those considerations, and these considerations being standardized between all motors in order to simplify maintenance.

rhfogh replied:
Good. Again, it is better with a simpler solution than standardising and making it all more difficult. Thanks for your patient explanations.

rhfogh commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSMotor.py:
```diff
@@ -0,0 +1,97 @@
+        super().abort()
+
+    def get_limits(self):
+        try:
```
Do the limits vary over time? If not, why not use the existing machinery?

beteva replied:
I believe this is the concrete implementation for getting the limits.

pedrobenetton replied:
Since the function call was not receiving anything from the `return`, I changed it from `get_limits()` to `set_limits()` and I think it's being used correctly now.

beteva replied:
Well, set and get do not have the same usage. set_limits is needed if you want to change the limits and not use the initially configured ones. You do not want to set limits if you only want to read them. I think your initial implementation of get_limits was correct. You were probably missing the concrete reading from the epics configuration if you did not see anything returned.
You still need get_limits method with reading of the epics configuration. Otherwise you will get (None, None), as this is the default value of self._nominal_limits if you do get_limits before set_limits.
You may want to check how get_limits is implemented in BlissMotor or ExporterMotor

pedrobenetton replied:
I did read BlissMotor and ExporterMotor and it seems that I should just revert this last commit. What I meant about the `return` was that `get_limits()` gets called at `init()` and doesn't do anything with what is returned, the returned values are correct.

rhfogh commented on mxcubecore/HardwareObjects/LNLS/LNLSDiffractometer.py:
```diff
@@ -0,0 +1,345 @@
+
+        GenericDiffractometer.init(self)
+        # Bzoom: 1.86 um/pixel (or 0.00186 mm/pixel) at minimum zoom
+        self.x_calib = 0.00186
```
This should be in a configuration file rather than hardcoded, on general principles. Or if you really, really, have to hardcode it, at least put it in a named constant at the top of the file.

pedrobenetton replied:
I have moved the values to the configuration file.

rhfogh reviewed (changes requested):
See comments to code

beteva commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,70 @@
+    ACTUATOR_RBV = "rbv"  # readback
+
+    def __init__(self, name):
+        super(EPICSActuator, self).__init__(name)
```
```suggestion
        super().__init__(name)
```

beteva replied:
I do not think it is resolved.

beteva commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,70 @@
+        if not self.unit:
+            self.unit = 10**-3
+
+    def init(self):
```
```suggestion
    def init(self):
        super().init()
```
Is it for purpose you are initialising the parent class?

beteva commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,70 @@
+            return False
+        return not np.isclose(readback, setpoint, rtol=self.unit, atol=self.unit)
+
+    def wait_ready(self, timeout):
```
```suggestion
    def wait_ready(self, timeout=None):
```

beteva commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSActuator.py:
```diff
@@ -0,0 +1,70 @@
+        self.set_channel_value(self.ACTUATOR_VAL, value)
+
+    def set_value(self, value, timeout: float = 0):
+        if not timeout:
+            timeout = self.default_timeout
```
You are excluding the default value (timeout=0) - is it deliberate? This also means that you cannot set timeout to None, which is the default behaviour for the rest of the systems.

pedrobenetton replied:
We did this because of the `AbstractActuator` logic at the `set_value` function:
If there is no timeout, `wait_ready` is never called, and then the `update_state` never happens, and the motor never returns to the `STATES.READY` state. After a few discussions it was clear that 180 seconds was more than enough for anything in the beamline for a `default_timeout`.

beteva replied:
> If there is no timeout, `wait_ready` is never called

This is only true if timeout=0. If timeout=None, it waits forever. In your code you exclude both 0 and None. That means:
if no 0 - you do not have the possibility to run two actions together.
if no None - you have to be sure that your default value is big enough for every set_value call or you specify appropriate timeout every time you do set_value.
This is up to you.

rhfogh replied:
IMHO this is bad - you are breaking the behaviour of the superclass. It says in the relevant documentation that if timeout=0 do not wait, and if timeout==None wait forever. Anyone calling that function is entitled to rely on that behaviour. If it is the case that the state  never returns to READY if the timeout is 0 then that is a bug ino the rest of the code, and that should be fixed, not worked around. Anyway, I wonder. Surely the actuator will return to ready when the motor has finished moving and sends the relevant signal??

RafaelLyra8 replied:
I did this because I always wanted to call the wait_ready method of the EPICSMotor, because the timeout is dynamically set there, and the only way I could do that, without changing the Abstract classes, was setting the timeout. Another way that could be done would be adding a new parameter to the set_value in the AbstractActuator with a default value that allows sub classes to set the timeout dynamically.

rhfogh replied:
OK, it sounds like this solution does what is needed, and we should certainly not make everything more complicated just to be standardising. I'll stop being obstructive ;-)

beteva commented on mxcubecore/HardwareObjects/LNLS/EPICS/EPICSMotor.py:
```diff
@@ -0,0 +1,97 @@
+        self.get_precision()
+        super().init()
+
+    def wait_ready(self, timeout):
```
```suggestion
    def wait_ready(self, timeout=None):
```

beteva reviewed (commented):
I presume the diffractometer is a temporary solution, so would not comment on it.

beteva reviewed (changes requested):
Could you, please, apply the requested changes and not mark them as resolved without being changed. Thank you

pedrobenetton commented:
> Could you, please, apply the requested changes and not mark them as resolved without being changed. Thank you

Hi, I didn't mark them as resolved. Currently I updated this branch with the most recent commits from mxcube:develop and now I'm looking at everything that has been commented.

beteva commented:
Welcome to the wonderful world of rebasing 😄
I believe you've made a slight mistake, as there are MAXIV commits in your branch.

beteva commented on mxcubecore/HardwareObjects/LNLS/LNLSDiffractometer.py:
```diff
@@ -0,0 +1,345 @@
+        # Armazena clique do usuário
+        self.last_centred_position[0] = x
+        self.last_centred_position[1] = y
+        print(f"Última posição clicada pelo usuário: x={x}, y={y}")
```
This is not resolved. Please, see the @marcus-oscarsson comment.

RafaelLyra8 replied:
This was an old version of the LNLSDiffractometer that will be refactored completely in another PR, I didn't include it in this one because it would become really big.

rhfogh reviewed (approved):
Lots of good work here. Thanks for your long answers to my comments.

RafaelLyra8 committed:
Add motor acceleration to the EpicsMotor timeout

RafaelLyra8 committed:
Reduce sleep for wait_actuator

RafaelLyra8 committed:
Fix linting

RafaelLyra8 committed:
Add v1 for EpicsActuator and EpicsMotor

RafaelLyra8 committed:
Improve wait ready methods for EPICSMotor and EPICSActuator

RafaelLyra8 committed:
Add LNLSDiffractometer

RafaelLyra8 committed:
Don't use motor_name so as to simplify the motor yml

RafaelLyra8 committed:
Fix monitoring PV motion during change value

RafaelLyra8 committed:
Remove docstrings

RafaelLyra8 committed:
Configure default RBV polling period for EPICSMotors

RafaelLyra8 committed:
Update PVs using the EpicsChannel class

RafaelLyra8 committed:
Remove unused Diffractometer methods

RafaelLyra8 committed:
Fix missing parameter on wait_actuator method override

RafaelLyra8 committed:
Fix motor has finished unit problem

RafaelLyra8 committed:
Remove the watcher from the EPICSActuator

The value will be updated using the EPICSCommand class.

RafaelLyra8 committed:
Instantiate Epics Motor attributes in a new function

RafaelLyra8 committed:
Add Base LNLSDiffractometer

RafaelLyra8 committed:
Use appropriate method to get an specific channel object

RafaelLyra8 committed:
Add EPICSMotor timeout

RafaelLyra8 committed:
Use print_log for notifying errors

RafaelLyra8 committed:
Use gevent.Timeout for wait actuator in EPICSActuator

Commit:
Use precision value correctly from EPICS PV

RafaelLyra8 committed:
Remove validation if the PV value has changed

This was done because the validation already occurs in the update_value function

Commit:
Fix linting

Commit:
Call super().__init__ correctly at EPICSActuator

Commit:
Revert "Remove default_timeout"

This reverts commit 04e7aa7ca34412e62653cb8010bf348a500848d7.

Commit:
Fix print message for actuator limits

Commit:
Remove default_timeout

If a movement does not have a timeout, it should be set to None.
This removes the necessity of a default_timeout variable, for
cases where the timeout is unknown.

Commit:
Use _ready_event to wait for movements at EPICSMotor and EPICSActuator

The function wait_ready is redefined in EPICSMotor because it uses
not only the hasnt_arrived() (from EPICSActuator), but also the
done_movement(), for verifying if the movement has ended.

Commit:
Move hard-coded values to configuration file

Commit:
Make timeout optional at wait_ready functions

Commit:
Try to use _ready_event at EPICSMotor wait_ready function

Commit:
Replace print() with self.print_log() at LNLSDiffractometer
